Add LeakSanitizer support in report-converter

**Describe the solution you would like**
Add support for LeakSanitizer reports in report-converter, so running with -t lsan finds and converts the leaks to plists. Lsan is described in more detail here: https://clang.llvm.org/docs/LeakSanitizer.html

**Additional context**
We're using codechecker to sanitize our codebase, and many of the issues are detected by LeakSanitizer, but we currently can't convert them to codechecker reports (report-converter does not find the LeakSanitizer entries in the logs)

Hey @csordasmarton 

Sorry for being late on this issue. I started developing the report-converter, I am attaching the git branch here as well. But when I run `make package` in `tools/report-converter` the following error is displayed:
```sh
# Install package in 'development mode'.
python3 -m pip install -e . -e ../codechecker_report_hash/
Obtaining file:///home/jay/codechecker/tools/report-converter
Obtaining file:///home/jay/codechecker/tools/codechecker_report_hash
ERROR: Could not find a version that satisfies the requirement codechecker_report_hash (from report-converter==0.1.0) (from versions: none)
ERROR: No matching distribution found for codechecker_report_hash (from report-converter==0.1.0)
make: *** [Makefile:43: package] Error 1
```

Can you please tell me what is going wrong here?

Github Branch: https://github.com/jay24rajput/codechecker/tree/leak_sanitizer

--- a/docs/supported_code_analyzers.md
+++ b/docs/supported_code_analyzers.md
@@ -45,7 +45,7 @@ CodeChecker result directory which can be stored to a CodeChecker server.
 | [MemorySanitizer](https://clang.llvm.org/docs/MemorySanitizer.html)    | ✓ |
 | [UndefinedBehaviorSanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html)    | ✓ |
 | [DataFlowSanitizer](https://clang.llvm.org/docs/DataFlowSanitizer.html)    | ✗ |
-| [LeakSanitizer](https://clang.llvm.org/docs/LeakSanitizer.html)    | ✗ |
+| [LeakSanitizer](https://clang.llvm.org/docs/LeakSanitizer.html)    | ✓ |
 
 We support to convert multiple sanitizer output to a CodeChecker report
 directory which can be stored to a CodeChecker server by using our

--- a/docs/tools/report-converter.md
+++ b/docs/tools/report-converter.md
@@ -11,6 +11,7 @@ a CodeChecker server.
   * [Address Sanitizer](#address-sanitizer)
   * [Memory Sanitizer](#memory-sanitizer)
   * [Thread Sanitizer](#thread-sanitizer)
+  * [Leak Sanitizer](#leak-sanitizer)
 * [Cppcheck](#cppcheck)
 * [Spotbugs](#spotbugs)
 * [Facebook Infer](#facebook-infer)
@@ -62,9 +63,9 @@ optional arguments:
   -t TYPE, --type TYPE  Specify the format of the code analyzer output.
                         Currently supported output types are: asan, clang-
                         tidy, coccinelle, cppcheck, cpplint, eslint,
-                        fbinfer, golint, kernel-doc, mdl, msan, pyflakes,
-                        pylint, smatch, sparse, sphinx, spotbugs, tsan,
-                        tslint, ubsan.
+                        fbinfer, golint, kernel-doc, lsan, mdl, msan, 
+                        pyflakes, pylint, smatch, sparse, sphinx, spotbugs, 
+                        tsan, tslint, ubsan.
   --meta [META [META ...]]
                         Metadata information which will be stored alongside
                         the run when the created report directory will be
@@ -97,6 +98,7 @@ Supported analyzers:
   fbinfer - Facebook Infer, https://fbinfer.com
   golint - Golint, https://github.com/golang/lint
   kernel-doc - Kernel-Doc, https://github.com/torvalds/linux/blob/master/scripts/kernel-doc
+  lsan - LeakSanitizer, https://clang.llvm.org/docs/LeakSanitizer.html
   mdl - Markdownlint, https://github.com/markdownlint/markdownlint
   msan - MemorySanitizer, https://clang.llvm.org/docs/MemorySanitizer.html
   pyflakes - Pyflakes, https://github.com/PyCQA/pyflakes
@@ -184,6 +186,19 @@ clang++ -fsanitize=thread -g tsan.cpp
 
 # Generate plist files from the output.
 report-converter -t tsan -o ./tsan_results tsan.output
+```
+
+### [Leak Sanitizer](https://clang.llvm.org/docs/LeakSanitizer.html)
+- Compile with `-g` and `-fsanitize=address`  to get proper debug information in your binary.
+```sh
+# Compile your program.
+clang -fsanitize=address -g lsan.c
+
+# Run your program and redirect the output to a file.
+ASAN_OPTIONS=detect_leaks=1 ./a.out > lsan.output 2>&1
+
+# Generate plist files from the output.
+report-converter -t lsan ./lsan_results lsan.output
 ```
 
 ## [Cppcheck](http://cppcheck.sourceforge.net/)

--- a/tools/report-converter/codechecker_report_converter/cli.py
+++ b/tools/report-converter/codechecker_report_converter/cli.py
@@ -64,6 +64,8 @@ from codechecker_report_converter.sparse.analyzer_result import \
     SparseAnalyzerResult  # noqa
 from codechecker_report_converter.cpplint.analyzer_result import \
     CpplintAnalyzerResult  # noqa
+from codechecker_report_converter.sanitizers.leak.analyzer_result import \
+    LSANAnalyzerResult  # noqa
 
 
 LOG = logging.getLogger('ReportConverter')
@@ -105,6 +107,7 @@ supported_converters = {
     SphinxAnalyzerResult.TOOL_NAME: SphinxAnalyzerResult,
     SparseAnalyzerResult.TOOL_NAME: SparseAnalyzerResult,
     CpplintAnalyzerResult.TOOL_NAME: CpplintAnalyzerResult,
+    LSANAnalyzerResult.TOOL_NAME: LSANAnalyzerResult
 }
 
 supported_metadata_keys = ["analyzer_command", "analyzer_version"]

--- a/tools/report-converter/codechecker_report_converter/sanitizers/leak/__init__.py
+++ b/tools/report-converter/codechecker_report_converter/sanitizers/leak/__init__.py
@@ -1,0 +1,7 @@
+# -------------------------------------------------------------------------
+#
+#  Part of the CodeChecker project, under the Apache License v2.0 with
+#  LLVM Exceptions. See LICENSE for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# -------------------------------------------------------------------------

--- a/tools/report-converter/codechecker_report_converter/sanitizers/leak/analyzer_result.py
+++ b/tools/report-converter/codechecker_report_converter/sanitizers/leak/analyzer_result.py
@@ -1,0 +1,37 @@
+# -------------------------------------------------------------------------
+#
+#  Part of the CodeChecker project, under the Apache License v2.0 with
+#  LLVM Exceptions. See LICENSE for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# -------------------------------------------------------------------------
+
+
+from codechecker_report_converter.analyzer_result import AnalyzerResult
+from codechecker_report_converter.plist_converter import PlistConverter
+
+from .output_parser import LSANParser
+
+
+class LSANAnalyzerResult(AnalyzerResult):
+    """ Transform analyzer result of Clang LeakSanitizer. """
+
+    TOOL_NAME = 'lsan'
+    NAME = 'LeakSanitizer'
+    URL = 'https://clang.llvm.org/docs/LeakSanitizer.html'
+
+    def parse(self, analyzer_result):
+        """ Creates plist files from the given analyzer result to the given
+        output directory.
+        """
+        parser = LSANParser()
+
+        content = self._get_analyzer_result_file_content(analyzer_result)
+        if not content:
+            return
+
+        messages = parser.parse_messages(content)
+
+        plist_converter = PlistConverter(self.TOOL_NAME)
+        plist_converter.add_messages(messages)
+        return plist_converter.get_plist_results()

--- a/tools/report-converter/codechecker_report_converter/sanitizers/leak/output_parser.py
+++ b/tools/report-converter/codechecker_report_converter/sanitizers/leak/output_parser.py
@@ -1,0 +1,55 @@
+# -------------------------------------------------------------------------
+#
+#  Part of the CodeChecker project, under the Apache License v2.0 with
+#  LLVM Exceptions. See LICENSE for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# -------------------------------------------------------------------------
+
+
+import logging
+import re
+
+from ...output_parser import get_next, Message, Event
+from ..output_parser import SANParser
+
+LOG = logging.getLogger('ReportConverter')
+
+
+class LSANParser(SANParser):
+    """ Parser for Clang LeakSanitizer console outputs. """
+
+    def __init__(self):
+        super(LSANParser, self).__init__()
+
+        # Regex for parsing MemorySanitizer output message.
+        self.leak_line_re = re.compile(
+            # Error code
+            r'==(?P<code>\d+)==(ERROR|WARNING): LeakSanitizer: '
+            # Checker message.
+            r'(?P<message>[\S \t]+)')
+
+    def parse_sanitizer_message(self, it, line):
+        """ Parses LeakSanitizer output message.
+
+        The first event will be the main location of the bug.
+        """
+        match = self.leak_line_re.match(line)
+        if not match:
+            return None, line
+
+        line = get_next(it)
+        stack_traces, events, line = self.parse_stack_trace(it, line)
+
+        if not events:
+            return None, line
+
+        main_event = events[-1]
+
+        notes = [Event(main_event.path, main_event.line, main_event.column,
+                       ''.join(stack_traces))]
+
+        return Message(main_event.path, main_event.line, main_event.column,
+                       match.group('message').strip(),
+                       "LeakSanitizer",
+                       events, notes), line

--- a/tools/report-converter/tests/unit/lsan_output_test_files/files/lsan.c
+++ b/tools/report-converter/tests/unit/lsan_output_test_files/files/lsan.c
@@ -1,0 +1,8 @@
+#include <stdlib.h>
+void *p;
+int main()
+{
+    p = malloc(7);
+    p = 0; // The memory is leaked here.
+    return 0;
+}

--- a/tools/report-converter/tests/unit/lsan_output_test_files/lsan.out
+++ b/tools/report-converter/tests/unit/lsan_output_test_files/lsan.out
@@ -1,0 +1,6 @@
+==23646==ERROR: LeakSanitizer: detected memory leaks
+Direct leak of 7 byte(s) in 1 object(s) allocated from:
+    #0 0x4af01b in __interceptor_malloc /projects/compiler-rt/lib/asan/asan_malloc_linux.cc:52:3
+    #1 0x4da26a in main files/lsan.c:4:7
+    #2 0x7f076fd9cec4 in __libc_start_main libc-start.c:287
+SUMMARY: AddressSanitizer: 7 byte(s) leaked in 1 allocation(s)

--- a/tools/report-converter/tests/unit/lsan_output_test_files/lsan.plist
+++ b/tools/report-converter/tests/unit/lsan_output_test_files/lsan.plist
@@ -1,0 +1,111 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>diagnostics</key>
+	<array>
+		<dict>
+			<key>category</key>
+			<string>unknown</string>
+			<key>check_name</key>
+			<string>LeakSanitizer</string>
+			<key>description</key>
+			<string>detected memory leaks</string>
+			<key>issue_hash_content_of_line_in_context</key>
+			<string>37e6fc6c7fb792d5e91ac6c3dbbe46b0</string>
+			<key>location</key>
+			<dict>
+				<key>col</key>
+				<integer>7</integer>
+				<key>file</key>
+				<integer>0</integer>
+				<key>line</key>
+				<integer>4</integer>
+			</dict>
+			<key>notes</key>
+			<array>
+				<dict>
+					<key>depth</key>
+					<integer>0</integer>
+					<key>kind</key>
+					<string>note</string>
+					<key>location</key>
+					<dict>
+						<key>col</key>
+						<integer>7</integer>
+						<key>file</key>
+						<integer>0</integer>
+						<key>line</key>
+						<integer>4</integer>
+					</dict>
+					<key>message</key>
+					<string>Direct leak of 7 byte(s) in 1 object(s) allocated from:
+    #0 0x4af01b in __interceptor_malloc /projects/compiler-rt/lib/asan/asan_malloc_linux.cc:52:3
+    #1 0x4da26a in main files/lsan.c:4:7
+    #2 0x7f076fd9cec4 in __libc_start_main libc-start.c:287
+SUMMARY: AddressSanitizer: 7 byte(s) leaked in 1 allocation(s)
+</string>
+				</dict>
+			</array>
+			<key>path</key>
+			<array>
+				<dict>
+					<key>depth</key>
+					<integer>0</integer>
+					<key>kind</key>
+					<string>event</string>
+					<key>location</key>
+					<dict>
+						<key>col</key>
+						<integer>7</integer>
+						<key>file</key>
+						<integer>0</integer>
+						<key>line</key>
+						<integer>4</integer>
+					</dict>
+					<key>message</key>
+					<string>    #1 0x4da26a in main files/lsan.c:4:7</string>
+				</dict>
+				<dict>
+					<key>depth</key>
+					<integer>0</integer>
+					<key>kind</key>
+					<string>event</string>
+					<key>location</key>
+					<dict>
+						<key>col</key>
+						<integer>7</integer>
+						<key>file</key>
+						<integer>0</integer>
+						<key>line</key>
+						<integer>4</integer>
+					</dict>
+					<key>message</key>
+					<string>detected memory leaks</string>
+				</dict>
+			</array>
+			<key>type</key>
+			<string>lsan</string>
+		</dict>
+	</array>
+	<key>files</key>
+	<array>
+		<string>files/lsan.c</string>
+	</array>
+	<key>metadata</key>
+	<dict>
+		<key>analyzer</key>
+		<dict>
+			<key>name</key>
+			<string>lsan</string>
+		</dict>
+		<key>generated_by</key>
+		<dict>
+			<key>name</key>
+			<string>report-converter</string>
+			<key>version</key>
+			<string>x.y.z</string>
+		</dict>
+	</dict>
+</dict>
+</plist>

--- a/tools/report-converter/tests/unit/test_lsan_parser.py
+++ b/tools/report-converter/tests/unit/test_lsan_parser.py
@@ -1,0 +1,115 @@
+# -------------------------------------------------------------------------
+#
+#  Part of the CodeChecker project, under the Apache License v2.0 with
+#  LLVM Exceptions. See LICENSE for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# -------------------------------------------------------------------------
+
+"""
+This module tests the correctness of the OutputParser and PListConverter, which
+used in sequence transform LeakSanitizer output to a plist file.
+"""
+
+
+import os
+import plistlib
+import shutil
+import tempfile
+import unittest
+
+from codechecker_report_converter.output_parser import Event, Message
+from codechecker_report_converter.sanitizers.leak.output_parser import \
+    LSANParser
+from codechecker_report_converter.sanitizers.leak.analyzer_result import \
+    LSANAnalyzerResult
+
+OLD_PWD = None
+
+
+def setup_module():
+    """Setup the test tidy reprs for the test classes in the module."""
+    global OLD_PWD
+    OLD_PWD = os.getcwd()
+    os.chdir(os.path.join(os.path.dirname(__file__),
+                          'lsan_output_test_files'))
+
+
+def teardown_module():
+    """Restore environment after tests have ran."""
+    global OLD_PWD
+    os.chdir(OLD_PWD)
+
+
+class LSANPListConverterTestCase(unittest.TestCase):
+    """ Test the output of the LSANAnalyzerResult. """
+
+    def setUp(self):
+        """ Setup the test. """
+        self.analyzer_result = LSANAnalyzerResult()
+        self.cc_result_dir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        """ Clean temporary directory. """
+        shutil.rmtree(self.cc_result_dir)
+
+    def test_san(self):
+        """ Test for the lsan.plist file. """
+        self.analyzer_result.transform('lsan.out', self.cc_result_dir)
+
+        with open('lsan.plist', mode='rb') as pfile:
+            exp = plistlib.load(pfile)
+
+        plist_file = os.path.join(self.cc_result_dir, 'lsan.c_lsan.plist')
+        with open(plist_file, mode='rb') as pfile:
+            res = plistlib.load(pfile)
+
+            # Use relative path for this test.
+            res['files'][0] = 'files/lsan.c'
+
+            self.assertTrue(res['metadata']['generated_by']['version'])
+            res['metadata']['generated_by']['version'] = "x.y.z"
+
+        self.assertEqual(res, exp)
+
+
+class LSANOutputParserTestCase(unittest.TestCase):
+    """
+    Tests the output of the OutputParser, which converts an Leak Sanitizer
+    output file to zero or more Message object.
+    """
+
+    def setUp(self):
+        """ Setup the OutputParser. """
+        self.parser = LSANParser()
+        self.lsan_repr = [
+            Message(
+                os.path.abspath('files/lsan.c'),
+                4, 7,
+                "detected memory leaks",
+                "LeakSanitizer",
+                [Event(
+                    os.path.abspath('files/lsan.c'),
+                    4, 7,
+                    "    #1 0x4da26a in main files/lsan.c:4:7"
+                )],
+                [Event(
+                    os.path.abspath('files/lsan.c'),
+                    4, 7,
+                    "Direct leak of 7 byte(s) in 1 object(s) allocated from:\n"
+                    "    #0 0x4af01b in __interceptor_malloc /projects/"
+                    "compiler-rt/lib/asan/asan_malloc_linux.cc:52:3\n"
+                    "    #1 0x4da26a in main files/lsan.c:4:7\n"
+                    "    #2 0x7f076fd9cec4 in __libc_start_main "
+                    "libc-start.c:287\n"
+                    "SUMMARY: AddressSanitizer: 7 byte(s) "
+                    "leaked in 1 allocation(s)\n"
+                )]),
+        ]
+
+    def test_lsan(self):
+        """ Test the generated Messages of lsan.out. """
+        messages = self.parser.parse_messages_from_file('lsan.out')
+        self.assertEqual(len(messages), len(self.lsan_repr))
+        for message in messages:
+            self.assertIn(message, self.lsan_repr)

